### ISSUE TITLE
``erase()`` function that takes an iterator but does not return one.

### ISSUE THREAD

**Issue by wjakob** [Tessil/robin-map#75]
The following program structure can lead to a quadratic computational cost as a function of `N`.
```cpp
tls::robin_map<int, int> map;
map.min_load_factor(.1);

// Insert N elements into a map
for (int i = 0; i < N; ++i)
    map[i] = i;

for (int i = 0; i < N; ++i) {
    auto it = map.find(i);
    if (it != map.end()) {
        map.erase(it);
    }
}
```

The problem is that `robin_map::erase` returns an iterator. To create a valid return value iterator, the implementation walks to the next bucket if the current one is empty, which can take a long time if the map is nearly empty. It also sets a hint that re-hashing may be in order, but it doesn't re-hash the map itself even if the load factor drops below the threshold.

Here is the code from `robin_hash.h`
 
```cpp
  iterator erase(iterator pos) {
    erase_from_bucket(pos);

    /**
     * Erase bucket used a backward shift after clearing the bucket.
     * Check if there is a new value in the bucket, if not get the next
     * non-empty.
     */
    if (pos.m_bucket->empty()) {
      ++pos;       // <<----------------- THIS IS THE EXPENSIVE STEP
    }

    m_try_shrink_on_next_insert = true;

    return pos;
  }
```

The example above is of course somewhat contrived, but in real applications, there are often usage patterns involving batch-insertion and batch-erasure of many elements. The batch-erasure can then easily suffer from this `O(n^2)` cost. A user of [nanobind](https://nanobind.readthedocs.io/en/latest/) reported an [issue](https://github.com/wjakob/nanobind/issues/507) in which they were able to trigger this performance issue. I worked around the problem but am not happy with the fix, hence the ticket here.

What I would really like to do is to call `erase_from_bucket` on the underlying `m_ht` (`tsl::robin_hash`) member, but this function is private and I don't see how to get to it. An other option is to call `size_t erase(K key)` or `size_t erase(K key, size_t hash)`, which don't suffer from the problem because they do not return an iterator. This works but is inefficient when we _already have an iterator_ (for example, because the application using `robin_map` needed to examine the value before deciding to erase it). The `size_t erase(K key, size_t hash)` variant then does redundant work by calling `find_impl` a second time to walk through neighboring buckets.

So this is a long-winded way of asking:

- Would you consider the addition of an `erase(iterator)` function that _does not_ return an iterator? I am happy to add the PR for it.
- Alternatively, would you consider adding a way to access the `m_ht` implementation detail?

**Comment by Tessil** [comment]
Effectively, adding an extra `void erase(iterator)` (though with another name to avoid conflict) would be a good addition.

I can check to work on it this weekend, but feel free to raise a PR if you have some time to work on it beforehand.

**Comment by wjakob** [comment]
> Effectively, adding an extra void erase(iterator) (though with another name to avoid conflict) would be a good addition.

Great, I am happy to hear it. I pushed a tentative fix in #76, please feel free to suggest another name. (I called it `erase_it` for now).
